action.php

```diff
@@ -187,30 +189,137 @@ public function handle_parser_metadata_render(Doku_Event $event, $param)
         }
     }
 
+    /**
+     * Returns JSON with filtered users and groups
+     *
+     * @param Doku_Event $event
+     * @param string $param
+     */
+    public function handle_ajax_get(Doku_Event $event, $param)
+    {
+        if ($event->data != 'plugin_watchcycle_get') return;
+        $event->preventDefault();
+        $event->stopPropagation();
+        global $conf;
+
+        header('Content-Type: application/json');
+        try {
+            $result = $this->fetchUsersAndGroups();
+        } catch(\Exception $e) {
+            $result = [
+                'error' => $e->getMessage().' '.basename($e->getFile()).':'.$e->getLine()
+            ];
+            if($conf['allowdebug']) {
+                $result['stacktrace'] = $e->getTraceAsString();
+            }
+            http_status(500);
+        }
+
+        $json = new JSON;
+        echo $json->encode($result);
+    }
+
+    /**
+     * JSON result of validation of maintainers definition
+     *
+     * @param Doku_Event $event
+     * @param $param
+     */
+    public function handle_ajax_validate(Doku_Event $event, $param)
+    {
+        if ($event->data != 'plugin_watchcycle_validate') return;
+        $event->preventDefault();
+        $event->stopPropagation();
+
+        global $INPUT;
+        $maintainers = $INPUT->str('param');
+
+        if (empty($maintainers)) return;
+
+        header('Content-Type: application/json');
+        $json = new JSON;
+
+        /* @var \helper_plugin_watchcycle $helper */
+        $helper = plugin_load('helper', 'watchcycle');
+
+        echo $json->encode($helper->validateMaintainerString($maintainers));
+    }
+
+    /**
+     * Returns filtered users and groups, if supported by the current authentication
+     *
+     * @return array
+     */
+    protected function fetchUsersAndGroups()
+    {
+        global $INPUT;
+        $term = $INPUT->str('param');
+
+        if (empty($term)) return [];
+
+        /* @var DokuWiki_Auth_Plugin $auth */
+        global $auth;
+
+        $users = [];
+        $foundUsers = $auth->retrieveUsers(0, 50, ['user' => $term]);
+        if (!empty($foundUsers)) {
+            $users = array_map(function ($name, $user) use ($term) {
+                return ['label' => $user['name'] . " ($name)", 'value' => $name];
+            }, array_keys($foundUsers), $foundUsers);
+        }
+
+        $groups = [];
+
+        // check cache
+        $cachedGroups = new cache('retrievedGroups', '.txt');
+        if($cachedGroups->useCache(['age' => 30])) {
+            $foundGroups = unserialize($cachedGroups->retrieveCache());
+        } else {
+            $foundGroups = $auth->retrieveGroups(0, 10);
```

Comment:
This retrieves 10 groups only, the cache is independent of the filter, filtering happens after retrieving the groups.

This means if the group I'm typing is not part of the first 10 groups returned (eg, it's the 12th) I will never see it in the autocompletion.

Reply:
You are right, limiting the results at this point does not make sense.